caddyconfig/httpcaddyfile/httptype.go

```diff
@@ -282,39 +282,18 @@ func (ServerType) evaluateGlobalOptionsBlock(serverBlocks []serverBlock, options
 		var val interface{}
 		var err error
 		disp := caddyfile.NewDispenser(segment)
-		switch dir {
-		case "debug":
-			val = true
-		case "http_port":
-			val, err = parseOptHTTPPort(disp)
-		case "https_port":
-			val, err = parseOptHTTPSPort(disp)
-		case "default_sni":
-			val, err = parseOptSingleString(disp)
-		case "order":
-			val, err = parseOptOrder(disp)
-		case "experimental_http3":
-			val, err = parseOptExperimentalHTTP3(disp)
-		case "storage":
-			val, err = parseOptStorage(disp)
-		case "acme_ca", "acme_dns", "acme_ca_root":
-			val, err = parseOptSingleString(disp)
-		case "email":
-			val, err = parseOptSingleString(disp)
-		case "admin":
-			val, err = parseOptAdmin(disp)
-		case "on_demand_tls":
-			val, err = parseOptOnDemand(disp)
-		case "local_certs":
-			val = true
-		case "key_type":
-			val, err = parseOptSingleString(disp)
-		default:
-			return nil, fmt.Errorf("unrecognized parameter name: %s", dir)
+
+		dirFunc, ok := registeredGlobalOptions[dir]
+		if !ok {
+			tkn := segment[0]
+			return nil, fmt.Errorf("%s:%d: unrecognized global option: %s", tkn.File, tkn.Line, dir)
 		}
+
+		val, err = dirFunc(disp)
 		if err != nil {
-			return nil, fmt.Errorf("%s: %v", dir, err)
+			return nil, fmt.Errorf("parsing caddyfile tokens for '%s': %v", dir, err)
```

Comment:
Better idea to wrap `err` (now that [we can](https://golang.org/doc/go1.13#error_wrapping)!) 😉

```suggestion
			return nil, fmt.Errorf("parsing caddyfile tokens for '%s': %w", dir, err)
```

Reply:
(note that the GitHub formatting doesn't recognize `%w` as a valid printf token yet - that's why it shows up red)

Reply:
That's news to me! Still new in Go 😛

Reply:
@hairyhenderson Any particular reason to do this? It exposes error types to the exported API, potentially unintentionally -- we should be sure we want to do this, i.e. have a good reason before doing that, since once we export it, it'll be hard to change/remove it later.

Reply:
@mholt How do you mean?

From https://golang.org/pkg/fmt/#Errorf:

> If the format specifier includes a %w verb with an error operand, the returned error will implement an Unwrap method returning the operand. It is invalid to include more than one %w verb or to supply it with an operand that does not implement the error interface. The %w verb is otherwise a synonym for %v.

So if there's an unexported error type, it should be safe, no?

Reply:
/shrug Maybe, although I still question the need for this change until there's a need for it.

Reply:
Fair enough. I just think that it's important to wrap errors where possible, to make downstream handling simpler. In this case it's no big deal.

Reply:
So what's the verdict, do I revert the change?